google/resource_spanner_database.go

```diff
@@ -0,0 +1,239 @@
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+
+					if len(value) < 2 && len(value) > 30 {
+						errors = append(errors, fmt.Errorf(
+							"%q must be between 2 and 30 characters in length", k))
+					}
+					if !regexp.MustCompile("^[a-z0-9-]+$").MatchString(value) {
+						errors = append(errors, fmt.Errorf(
+							"%q can only contain lowercase letters, numbers and hyphens", k))
+					}
+					if !regexp.MustCompile("^[a-z]").MatchString(value) {
+						errors = append(errors, fmt.Errorf(
+							"%q must start with a letter", k))
+					}
+					if !regexp.MustCompile("[a-z0-9]$").MatchString(value) {
+						errors = append(errors, fmt.Errorf(
+							"%q must end with a number or a letter", k))
+					}
+					return
+				},
+			},
+
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"ddl": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceSpannerDatabaseCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	id, err := buildSpannerDatabaseId(d, config)
+	if err != nil {
+		return err
+	}
+
+	cdr := &spanner.CreateDatabaseRequest{}
+	cdr.CreateStatement = fmt.Sprintf("CREATE DATABASE `%s`", id.Database)
+	if v, ok := d.GetOk("ddl"); ok {
+		cdr.ExtraStatements = convertStringArr(v.([]interface{}))
+	}
+
+	op, err := config.clientSpanner.Projects.Instances.Databases.Create(
+		id.parentInstanceUri(), cdr).Do()
+	if err != nil {
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusConflict {
+			return fmt.Errorf("Error, A database with name %s already exists in this instance", id.Database)
+		}
+		return fmt.Errorf("Error, failed to create database %s: %s", id.Database, err)
+	}
+
+	d.SetId(id.terraformId())
+
+	// Wait until it's created
+	timeoutMins := int(d.Timeout(schema.TimeoutCreate).Minutes())
+	waitErr := spannerDatabaseOperationWait(config, op, "Creating Spanner database", timeoutMins)
+	if waitErr != nil {
+		// The resource didn't actually create
+		d.SetId("")
+		return waitErr
+	}
+
+	log.Printf("[INFO] Spanner database %s has been created", id.terraformId())
+	return resourceSpannerDatabaseRead(d, meta)
+}
+
+func resourceSpannerDatabaseRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	id, err := buildSpannerDatabaseId(d, config)
+	if err != nil {
+		return err
+	}
+
+	_, err = config.clientSpanner.Projects.Instances.Databases.Get(
+		id.databaseUri()).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Spanner database %q", id.databaseUri()))
+	}
+
+	d.SetId(id.terraformId())
```

Comment:
No need to set the id here, it is set either in the create method or the import method.